File: data_structure/queue.py
# Python has in-built class(in the library `collections`) for implementing a queue.
# use `from collections import deque` to import deque,
# although deque is generalistion of stack and queue, we can use it to implement queue
# use `<queue-name>.append(<what-has-to-be-appended>)` to enqueue in the queue and
# use `<queue-name>.popleft()` to dequeue from the queue.

class Element:
    def __init__(self,value):
        self.value=value
        self.next=None

class Queue:
    def __init__(self,first_element=None):
        self.head=first_element
        self.tail=first_element

    def enqueue(self,new_element):
        if self.tail:
            self.tail.next=new_element
            self.tail=new_element
        else:
            self.tail=new_element

    def dequeue(self):
        if self.head:
            temp=self.head.value
            self.head=self.head.next
            return temp
        else:
            print "No element to dequeue!!!"

    def peak(self):
        if self.head:
            return self.head.value
        else:
            print "No element to peak!!!"

e1=Element(35)
e2=Element(80)

my_queue=Queue(e1)

my_queue.enqueue(e2)

print my_queue.head.value

File: data_structure/queue.cpp
/*
	Queue implemented using Linked List
*/

#include <iostream>
using namespace std;

struct node{
	int value;
	node *next;
};

class queue{
	node *head;
	public:
		queue(){
			head=NULL;
		}
		void enqueue(int val){
			if(head==NULL){
				head=new node;
				head->value=val;
				head->next=NULL;
			}
			else{
				node *temp=head;
				while(temp->next != NULL)
					temp=temp->next;

				temp->next=new node;
				temp->next->value=val;
				temp->next->next=NULL;
			}
		}
		int dequeue(){
			if(head==NULL){
				cout<<"Nothing to dequeue!!"<<endl;
				return -1;
			}
			else{
				node *temp=head;
				int val=head->value;
				head=head->next;
				delete temp;
				return val;
			}
		}
		void print(){
			if(head==NULL){
				cout<<"Queue is empty!!"<<endl;
			}
			else{
				node *temp=head;
				while(temp != NULL)
				{
					cout << temp->value << "<-";
					temp=temp->next;
				}
				cout<<endl;
			}
		}

};

int main(){
	queue q;

	q.enqueue(3);
	q.enqueue(4);
        q.enqueue(8);
	q.enqueue(19);

	cout<<q.dequeue()<<endl;
	//outputs 3

	q.print();
	//outputs 4<-8<-19<-

	return 0;
}
